### PR TITLE
Update and re-enable the diagnostics for unstable runtime names

### PR DESCRIPTION
**Pull request by jrose-apple**
(builds on #10078)

This time, the warnings only fire when the class in question directly conforms to NSCoding. This avoids warning on cases where the user has subclassed something like, oh, UIViewController, and has no intention of writing it to a persistent file.

This also removes the warning for generic classes that conform to NSCoding, for simplicity's sake. That means `@NSKeyedArchiverEncodeNonGenericSubclassesOnly` is also being removed.

Actually archiving a class with an unstable mangled name is still considered problematic, but the compiler shouldn't emit diagnostics unless it can be sure they are relevant.

rdar://problem/32314195

### CODE REVIEW

**Comment by jrose-apple** [comment]
cc @amartini51

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -1481,10 +1481,36 @@ bool Parser::parseDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc) {
     return false;
   }
 
+  // FIXME: Remove this after Swift 4.
   if (DK == DAK_Count && Tok.getText() == "NSKeyedArchiverClassName") {
-    DK = DAK_ObjC;
-    diagnose(Tok, diag::attr_nskeyedarchiverclassname_removed)
-      .fixItReplace(Tok.getLoc(), "objc");
+    auto activeDiag = diagnose(Tok,diag::attr_nskeyedarchiverclassname_removed);
+    activeDiag.fixItReplace(Tok.getLoc(), "objc");
```
Comment (DougGregor) [inline lib/Parse/ParseDecl.cpp]: Ah, I see you corrected the Fix-It here that I complained about in the previous PR. Thanks!

**Review by DougGregor — APPROVED** [review]
LGTM. As with the previous PR, this is a much simpler scheme.

**Review by itaiferber — APPROVED** [review]
Looks reasonable to me! Thanks for putting this together!

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Pull request merged**